File: yyxarch/src/main/java/com/yyxnb/yyxarch/base/mvvm/BaseMvvmActivity.java
package com.yyxnb.yyxarch.base.mvvm;

import android.arch.lifecycle.LifecycleObserver;
import android.arch.lifecycle.ViewModelProviders;
import android.os.Bundle;
import android.support.annotation.NonNull;

import com.yyxnb.yyxarch.AppUtils;
import com.yyxnb.yyxarch.base.BaseActivity;

public abstract class BaseMvvmActivity<VM extends BaseViewModel> extends BaseActivity {

    /**
     * ViewModel
     */
    protected VM mViewModel;

    @Override
    public void initView(Bundle savedInstanceState) {
        super.initView(savedInstanceState);
        mViewModel = initViewModel(this, (AppUtils.getInstance(this, 0)));
        if (mViewModel != null) {
            getLifecycle().addObserver(mViewModel);
        }
    }

    /**
     * 初始化ViewModel
     * create ViewModelProviders
     *
     * @return ViewModel
     */
    private VM initViewModel(BaseActivity activity, @NonNull Class<VM> modelClass) {
        return ViewModelProviders.of(activity).get(modelClass);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        //移除LifecycleObserver
        if (mViewModel != null) {
            getLifecycle().removeObserver(mViewModel);
        }
        this.mViewModel = null;
    }
}

File: yyxarch/src/main/java/com/yyxnb/yyxarch/base/mvvm/BaseMvvmActivity.kt
package com.yyxnb.yyxarch.base.mvvm

import android.arch.lifecycle.ViewModelProviders
import android.os.Bundle
import android.support.v4.app.FragmentActivity
import com.yyxnb.yyxarch.AppUtils
import com.yyxnb.yyxarch.base.BaseActivity

abstract class BaseMvvmActivity<VM : BaseViewModel<*>> : BaseActivity() {

    /**
     * ViewModel
     */
    protected var mViewModel: VM? = null

    override fun initView(savedInstanceState: Bundle?) {
        super.initView(savedInstanceState)
        mViewModel = initViewModel(this, AppUtils.getInstance(this, 0)!!)
        if (mViewModel != null) {
            lifecycle.addObserver(mViewModel!!)
        }
    }

    /**
     * 初始化ViewModel
     * create ViewModelProviders
     *
     * @return ViewModel
     */
    private fun initViewModel(activity: FragmentActivity, modelClass: Class<VM>): VM {
        return ViewModelProviders.of(activity).get(modelClass)
    }

    override fun onDestroy() {
        super.onDestroy()
        //移除LifecycleObserver
        if (mViewModel != null) {
            lifecycle.removeObserver(mViewModel!!)
        }
        this.mViewModel = null
    }
}
